### ISSUE TITLE
Authenticate type in BasicAuthenticationRequested event

### ISSUE THREAD

**Issue by sayabokt** [MicrosoftEdge/WebView2Feedback#3170]
I want to get the authentication type (HTTP status 401 or 407) in BasicAuthenticationRequested event.
For 401 I want to authenticate with WebView2 and for 407 I want to authenticate with a custom UI.
CoreWebView2BasicAuthenticationRequestedEventArgs should be able to retrieve the authentication type.

I can get the authentication type by checking the WebErrorStatus in the NavigationCompleted event,
It does not work properly when accessing a website that goes through a proxy from the content of the website that does not go through the proxy.

An example is shown below.

Site through proxy: https://www.proxy.example/
Site without proxy: https://www.no-proxy.example/ (loads JS scripts from www.proxy.example)

- works correctly
1. NavigationStarting (www.proxy.example)
2. WebResourceRequested (www.proxy.example)
3. WebResourceResponseReceived(www.proxy.example, HTTP status = 407)
4. BasicAuthenticationRequested (www.proxy.example)
    -> Defer processing with GetDeferral
5. NavigationCompleted(WebErrorStatus = ValidProxyAuthenticationRequired)

- not working properly
1. NavigationStarting (www.no-proxy.example)
2. WebResourceRequested (www.no-proxy.example)
3. WebResourceResponseReceived (www.no-proxy.example, HTTP status = 200)
4. WebResourceRequested event (www.proxy.example/ext.js)
5. BasicAuthenticationRequested(www.proxy.example/ext.js, HTTP status = 407)
    -> Defer processing with GetDeferral
6. NavigationCompleted event is not fired

**Comment by nishitha-burman** [comment]
Hi @sayabokt,

Thank you for the feedback! Can you please provide more details on issues you are running into and what's not working when checking WebErrorStatus in the NavigationCompleted event?

Thanks!

**Comment by sayabokt** [comment]
I have a proxy server and a project that I can reproduce.
The project is based on WebView2WindowsFormsBrowser from MicrosoftEdge/WebView2Samples.

[WebView2WindowsFormsBrowser.zip](https://github.com/MicrosoftEdge/WebView2Feedback/files/10558542/WebView2WindowsFormsBrowser.zip)

**Base commit**
MicrosoftEdge/WebView2Samples@5e5c0f9e7bc262c85474453e22b82aa65f9be6c4

**Diff**
<details><summary>diff</summary><p>

```diff
 .../BrowserForm.Designer.cs                   |  2 +-
 .../BrowserForm.cs                            | 73 ++++++++++++++++---
 2 files changed, 62 insertions(+), 13 deletions(-)

diff --git a/SampleApps/WebView2WindowsFormsBrowser/BrowserForm.Designer.cs b/SampleApps/WebView2WindowsFormsBrowser/BrowserForm.Designer.cs
index c4e13ed..e93f65e 100644
--- a/SampleApps/WebView2WindowsFormsBrowser/BrowserForm.Designer.cs
+++ b/SampleApps/WebView2WindowsFormsBrowser/BrowserForm.Designer.cs
@@ -157,7 +157,7 @@ namespace WebView2WindowsFormsBrowser
             this.txtUrl.Name = "txtUrl";
             this.txtUrl.Size = new System.Drawing.Size(738, 31);
             this.txtUrl.TabIndex = 4;
-            this.txtUrl.Text = "https://www.bing.com/";
+            this.txtUrl.Text = "about:blank";
             //
             // menuStrip1
             //
diff --git a/SampleApps/WebView2WindowsFormsBrowser/BrowserForm.cs b/SampleApps/WebView2WindowsFormsBrowser/BrowserForm.cs
index 4d8f787..9e6c4d8 100644
--- a/SampleApps/WebView2WindowsFormsBrowser/BrowserForm.cs
+++ b/SampleApps/WebView2WindowsFormsBrowser/BrowserForm.cs
@@ -31,6 +31,32 @@ namespace WebView2WindowsFormsBrowser
             }
         }
 
+        private CoreWebView2BasicAuthenticationRequestedEventArgs _basicAuthArgs = null;
+        public CoreWebView2BasicAuthenticationRequestedEventArgs BasicAuthArgs
+        {
+            get
+            {
+                return _basicAuthArgs;
+            }
+            set
+            {
+                _basicAuthArgs = value;
+            }
+        }
+
+        private CoreWebView2Deferral _basicAuthDeferral = null;
+        public CoreWebView2Deferral BasicAuthDeferral
+        {
+            get
+            {
+                return _basicAuthDeferral;
+            }
+            set
+            {
+                _basicAuthDeferral = value;
+            }
+        }
+
         public BrowserForm()
         {
             InitializeComponent();
@@ -116,6 +142,33 @@ namespace WebView2WindowsFormsBrowser
 
         private void WebView2Control_NavigationCompleted(object sender, CoreWebView2NavigationCompletedEventArgs e)
         {
+            if (e.IsSuccess == false)
+            {
+                if (e.WebErrorStatus == CoreWebView2WebErrorStatus.ValidProxyAuthenticationRequired)
+                {
+                    UpdateTitleWithEvent("ValidProxyAuthenticationRequired");
+
+                    if (BasicAuthDeferral != null)
+                    {
+                        BasicAuthDeferral.Complete();
+
+                        BasicAuthArgs = null;
+                        BasicAuthDeferral = null;
+                    }
+                }
+                else if (e.WebErrorStatus == CoreWebView2WebErrorStatus.ValidAuthenticationCredentialsRequired)
+                {
+                    UpdateTitleWithEvent("ValidAuthenticationCredentialsRequired");
+
+                    BasicAuthDeferral.Complete();
+
+                    BasicAuthArgs = null;
+                    BasicAuthDeferral = null;
+                }
+
+                return;
+            }
+
             UpdateTitleWithEvent("NavigationCompleted");
         }
 
@@ -141,10 +194,17 @@ namespace WebView2WindowsFormsBrowser
             this.webView2Control.CoreWebView2.HistoryChanged += CoreWebView2_HistoryChanged;
             this.webView2Control.CoreWebView2.DocumentTitleChanged += CoreWebView2_DocumentTitleChanged;
             this.webView2Control.CoreWebView2.AddWebResourceRequestedFilter("*", CoreWebView2WebResourceContext.Image);
+            this.webView2Control.CoreWebView2.BasicAuthenticationRequested += CoreWebView2_BasicAuthenticationRequested;
             UpdateTitleWithEvent("CoreWebView2InitializationCompleted succeeded");
             EnableButtons();
         }
 
+        private void CoreWebView2_BasicAuthenticationRequested(object sender, CoreWebView2BasicAuthenticationRequestedEventArgs e)
+        {
+            BasicAuthArgs = e;
+            BasicAuthDeferral = e.GetDeferral();
+        }
+
         void AttachControlEventHandlers(Microsoft.Web.WebView2.WinForms.WebView2 control) {
             control.CoreWebView2InitializationCompleted += WebView2Control_CoreWebView2InitializationCompleted;
             control.NavigationStarting += WebView2Control_NavigationStarting;
@@ -643,18 +703,7 @@ namespace WebView2WindowsFormsBrowser
 
         private string GetStartPageUri(CoreWebView2 webView2)
         {
-            string uri = "https://appassets.example/AppStartPage.html";
-            if (webView2 == null)
-            {
-                return uri;
-            }
-            string sdkBuildVersion = GetSdkBuildVersion(),
-                   runtimeVersion = GetRuntimeVersion(webView2),
-                   appPath = GetAppPath(),
-                   runtimePath = GetRuntimePath(webView2);
-            string newUri = $"{uri}?sdkBuild={sdkBuildVersion}&runtimeVersion={runtimeVersion}" +
-                $"&appPath={appPath}&runtimePath={runtimePath}";
-            return newUri;
+            return "about:blank";
         }
     }
 }
```
</p></details>

**Proxy server**

squid v4.10
150.230.206.105:3128

require password: .bing.com
id: user
passowrd: !qazxsw2

<details><summary>squid.conf</summary><p>

```
auth_param basic program /usr/lib/squid/basic_ncsa_auth /etc/squid/.user.passwd
auth_param basic children 5 startup=5 idle=1
auth_param basic realm from proxy
auth_param basic credentialsttl 2 hours

acl password proxy_auth REQUIRED
icap_send_client_username on

acl bing dstdomain .bing.com
http_access allow bing
http_access allow password
http_access deny all
```
</p></details>

**Repro Steps**

Runtime: 109.0.1518.70
OS: Windows 10 22H2 (19045.2546)

1. build project.
2. run WebView2WindowsFormsBrowser.exe with option as below.
   WebView2WindowsFormsBrowser.exe --edge-webview-switches=--proxy-server=150.230.206.105:3128
3. load https://www.bing.com/
4. NavigationComplete does not fire when the below request occurs.
   https://assets.msn.com/bundles/v1/bingHomepage/latest/widget-initializer.js

**Screenshot**

![pending-request](https://user-images.githubusercontent.com/92766393/216085091-48a8b1f0-3065-4e26-9fc3-cf1543e6f779.png)

**Comment by sayabokt** [comment]
@nishitha-burman Any updates on this?

**Comment by novac42** [comment]
Hi @sayabokt we've assigned this to a dev that can help follow up on this.

**Comment by sayabokt** [comment]
> I want to get the authentication type (HTTP status 401 or 407) in BasicAuthenticationRequested event.

This request will be completely resolved, if #3250 fixed.
In the meantime, i will identify auth type with checking uri.
I will close this issue.
